buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -135,7 +135,12 @@ enum MagazineCaching {
                 protected Object initialValue() {
                     if (cachedMagazinesNonEventLoopThreads || ThreadExecutorMap.currentExecutor() != null) {
                         Magazine mag = new Magazine(AdaptivePoolingAllocator.this, false);
-                        liveMagazines.add(mag);
+
+                        if (FastThreadLocalThread.willCleanupFastThreadLocals(Thread.currentThread())) {
+                            // Only add it to the liveMagazines if we can guarantee that onRemoval(...) is called,
+                            // as otherwise we might end up holding the reference forever.
+                            liveMagazines.add(mag);
```

Comment:
The downside is that if we don't add it the `usedMemory()` method might not be 100 % accurate.

Reply:
If the fast thread local cannot get rid of the magazine strong reference. probably is not a good idea either to assign a magazine, too? Just thinking loud eh

Reply:
I am not sure I follow... It will GC'ed the same way as a normal ThreadLocal in any case. It's just that the onRemoval(...) might not be called.

Reply:
right, friday brain, forget this  comment :)